warp/aggregator/aggregator.go

```diff
@@ -36,16 +71,150 @@ func (a *Aggregator) AggregateSignatures(ctx context.Context, unsignedMessage *a
 	if err != nil {
 		return nil, err
 	}
-	job := newSignatureAggregationJob(
-		a.client,
-		pChainHeight,
-		a.subnetID,
-		quorumNum,
-		quorumNum,
-		params.WarpQuorumDenominator,
-		a.state,
-		unsignedMessage,
+
+	log.Debug("Fetching signature",
+		"a.subnetID", a.subnetID,
+		"height", pChainHeight,
 	)
+	validators, totalWeight, err := avalancheWarp.GetCanonicalValidatorSet(ctx, a.state, pChainHeight, a.subnetID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get validator set: %w", err)
+	}
+	if len(validators) == 0 {
+		return nil, fmt.Errorf("%w (SubnetID: %s, Height: %d)", errNoValidators, a.subnetID, pChainHeight)
+	}
+
+	return aggregateSignatures(ctx, a.client, unsignedMessage, validators, quorumNum, totalWeight)
+}
+
+func aggregateSignatures(
+	ctx context.Context,
+	client SignatureGetter,
+	unsignedMessage *avalancheWarp.UnsignedMessage,
+	validators []*avalancheWarp.Validator,
+	quorumNum uint64,
+	totalWeight uint64,
+) (*AggregateSignatureResult, error) {
+	type signatureFetchResult struct {
+		sig    *bls.Signature
+		index  int
+		weight uint64
+	}
+
+	// Create a child context to cancel signature fetching if we reach signature threshold.
+	signatureFetchCtx, signatureFetchCancel := context.WithCancel(ctx)
+	defer signatureFetchCancel()
+
+	// Fetch signatures from validators concurrently.
+	signatureFetchResultChan := make(chan *signatureFetchResult, len(validators))
```

Comment:
```suggestion
	signatureFetchResultChan := make(chan *signatureFetchResult)
```
I don't think this channel needs to be buffered?